message_ix/macro.py

```diff
@@ -193,18 +190,28 @@ def __init__(self, s, data):
         if not s.has_solution():
             raise RuntimeError("Scenario must have a solution to add MACRO")
 
-        demand = s.var("DEMAND", filters={"level": "useful"})
-        self.nodes = set(demand["node"].unique())
-        self.sectors = set(demand["commodity"].unique())
-        self.years = set(demand["year"].unique())
+        if 'config' not in self.data:
+            raise KeyError('Missing config in input data')
+        else:
+            config = self.data['config']
+            for key in ['node', 'sector', 'level']:
+                try:
+                    config[key]
+                except KeyError:
+                    raise KeyError('Missing config data for "{}"'.format(key))
+                else:
+                    if config[key].dropna().empty:
+                        raise ValueError(
+                            'Config data for "{}" is empty'.format(key))
+
+        demand = s.var('DEMAND')
+        self.years = set(demand['year'].unique())
 
     def read_data(self):
-        if "config" in self.data:
-            # users may remove certain nodes and sectors from the MACRO set
-            # definitions
-            config = self.data["config"]
-            self.nodes -= set(config.get("ignore_nodes", []))
-            self.sectors -= set(config.get("ignore_sectors", []))
+        # users define certain nodes, sectors and level for MACRO
+        self.nodes = set(self.data['config']['node'].dropna())
+        self.sectors = set(self.data['config']['sector'].dropna())
+        self.levels = set(self.data['config']['level'].dropna())
```

Comment:
Here we can also read the commodities from the configuration sheet to be later used in the "mapping_macro_sector"set:

`self.commodities = set(self.data['config']['commodity'].dropna())`